## Python Lab 03

In [1]:
import numpy as np

---

# Task
Numpy task

* Generate $N\times 2$ array of random numbers $[0,1]$
* Calculate squares of all elements
* Add elements along the short direction (resulting in $N\times 1$ array)
* Count the number of elements less than 1
* Calculate the value:
$$4\cdot\frac{N_{x<1}}{N}$$

In [47]:
N = 100000000
mat = np.random.random((N,2))

In [48]:
mat2=mat**2

In [49]:
mats=np.sum(mat2,axis=1)
mats

array([0.31382657, 0.79153118, 0.86384935, ..., 0.90892555, 1.36397125,
       0.13399536], shape=(100000000,))

In [50]:
# NumPy equivalent of: [i for i in _ if i>1]
# use existing 1D array `mats`
matx = mats[mats<1]

In [36]:
matx = [i for i in mats if i<1]

In [51]:
4*len(matx)/N

3.14171432

In [52]:
np.pi

3.141592653589793

almost there...

## Task
    
Implement a function calculating price $P$    

   $$P(y;T, \vec{C}, F)=\frac{F}{(1+y)^T}+\sum_{i=1}^T\frac{C_i}{(1+y)^i}$$
   
   
   From coupons $\vec{C}\in \mathbb{R}^T$ (narray) and given yield-to-maturity $y$ and face value $F$.
  For example choose the following values:
  
   $$y=5\%,\quad F=\$100, \quad C_i=\$3,\quad T=29$$

---


# Task

## 
Consider a logistic map $\scriptsize x\ni[0,1]\to f(x)=kx(1-x)\in[0,1]$ for $\scriptsize k\in[0,4]$.

The iteration 

$$\scriptsize x_{n+1}=kx_n(1-x_n)$$

has two fixed points (solution to $\scriptsize x=f(x)$):  $\scriptsize x=0$ and $\scriptsize x=1-\frac1k$.

## 
Q: What happens when $\scriptsize k=0.5$, $\scriptsize k=2.5$, $\scriptsize k=3.1$, $\scriptsize k=3.5$, $\scriptsize k=4$? And why?

More precisely:
* print the series $x_n$ for given $k$s
* plot the series in those cases
* Draw a graph showing $(k, x_n)$ for many numbers $n>100$ and different $k\in [0,4]$
[similar to this](https://en.wikipedia.org/wiki/Logistic_map#/media/File:Logistic_Bifurcation_map_High_Resolution.png)

can we speed up the process.

---

# Task
## 
find a solution to the equation:

$$x=\cos(x)$$


for $x\in [0,1]$ using three methods:
* bisection: <br>
    we divide an interval $[a, b]$, to $[a,c]$  and $[c, b]$ for $c=\frac{a+b}{2}$ and for next iteration we choose the interval where the sign is changed.
* naive iteration $x_{n+1}=\cos(x_n)$,
* Newton's method:<br>
 for $f(x)=0$ we iterate $$x_{n+1}=x_{n}-\frac{f(x_n)}{f'(x_n)}$$

Plot function $y=x$ and $y=\cos(x)$.<br>
Analyze how fast the solution is reached.

## 
Solve the equation 
$$P(y; \vec{C}, F) = 53.52$$

$$F(x)=\cos(x)-x$$

---

# Task
Analyze Newton's method for 

$$f(x)=|x|^p$$

for different values of $p$.

---

# Task
## 
Analyze the Newton's method for 

$$f(z)=z^3-1,\;\; f'(z)=3z^2$$

$$z=e^{2m\pi i/3},\;\; m=1,2,3$$

for complex valued $z$. Draw the Newton's fractal.<br>
How effectively can you do it?
## 
Draw the Hayley fractal

## 
Draw the Newton fractal for equation
$$f(z) = z^3-2z+2$$